### DISCUSSION TITLE
Controlling <Tab> Completion (Again)

### DISCUSSION THREAD

**Discussion by ewhac** [neoclide/coc.nvim#3378] [Q&A]
Here's what I want:  If a completion list popup is not being displayed, then I want the `<Tab>` key to insert indentation characters, be they spaces or a literal TAB character.

However, if a completion list popup is being displayed, then I want `<Tab>/<S-Tab>` to cycle forwards/backwards through the available items.

After reviewing the pertinent docs, I'm unable to figure out how to accomplish what I want.  I'm finding it increasingly frustrating to press `<Tab>` after entering a keyword/variable/function name expecting to advance the cursor, only to have the completion popup reappear.  I've tried setting `suggest.autoTrigger` to `trigger`, but that doesn't seem to have helped.

My thanks in advance.

**Comment by fannheyward** [comment]
Does this work for you?

```viml
inoremap <silent><expr> <TAB>
      \ pumvisible() ? "\<C-n>" :
      \ <SID>check_back_space() ? "\<TAB>" :
      \ coc#refresh()
```

**Reply by ewhac** [reply]
It does not; I already have exactly that in an init script.  In fact, that very snippet appears in `doc/coc.txt` describing how to trigger completion using TAB.

Hrm.  Staring at that snippet for a few minutes, it looks like what I want might be:

```vim
inoremap <silent><expr> <TAB>
	\ pumvisible() ? "\<C-n>" :
	\ "\<TAB>"
```

Hmm.  That's just about what I want.  Alas, it looks like nvim-qt doesn't handle \<C-space\> correctly yet, so I'll have to pick a different trigger key...
